pegasus_wrapper/workflow.py

```diff
@@ -407,7 +787,166 @@ def run_python_on_args(
+            job_profiles=job_profiles,
             treat_params_as_cmd_args=True,
+            input_file_paths=input_file_paths,
+            output_file_paths=output_file_paths,
+        )
+
+    def run_bash(
+        self,
+        job_name: Locator,
+        command: Union[Iterable[str], str],
+        *,
+        depends_on,
+        resource_request: Optional[ResourceRequest] = None,
+        category: Optional[str] = None,
+        job_is_stageable: bool = False,
+        job_bypass_staging: bool = False,
+        times_to_retry_job: int = 0,
+        job_profiles: Optional[List[PegasusProfile]] = None,
+        container: Optional[Container] = None,
+        path_to_bash: Path = Path("/usr/bin/bash"),
+    ) -> DependencyNode:
+        """
+        Schedule a job to run the given *command* with the given *resource_request*
+
+        If this job requires other jobs to be executed first,
+        include them in *depends_on*.
+
+        This method returns a `DependencyNode` which can be used in *depends_on*
+        for future jobs.
+        """
+        if job_profiles is None:
+            job_profiles = []
+
+        if isinstance(command, str):
+            command = [command]
+
+        commands_hashable = immutableset(command)
+
+        signature = (job_name, commands_hashable)
+        if signature in self._signature_to_job:
+            logging.info("Job %s recognized as duplicate", job_name)
+            return self._signature_to_job[signature]
+
+        bash_transform = self._define_transformation(
+            "bash",
+            str(path_to_bash.absolute()),
+            site=self._default_site,
+            container=container,
+            is_stageable=job_is_stageable,
+            bypass_staging=job_bypass_staging,
+        ).transformation
+
+        job_dir = self.directory_for(job_name)
+        ckpt_name = job_name / "___ckpt"
+        ckpt_path = job_dir / "___ckpt"
+        job_script = job_dir / "script.sh"
+
+        commands_with_ckpt = list(command)
+        commands_with_ckpt.append(f"touch {ckpt_path.absolute()}")
+
+        job_script.write_text("\n".join(commands_with_ckpt))
+
+        bash_job = Job(bash_transform)
+        bash_job.add_args(str(job_script.absolute()))
+        bash_job.add_dagman_profile(category=category, retry=str(times_to_retry_job))
+
+        for profile in job_profiles:
+            bash_job.add_profiles(profile.namespace, key=profile.key, value=profile.value)
+
+        resource_request = self.set_resource_request(resource_request)
+        resource_request.apply_to_job(bash_job, job_name=self._job_name_for(job_name))
+
+        self._job_graph.add_jobs(bash_job)
+        for parent_dependency in depends_on:
+            if parent_dependency.job:
+                self._job_graph.add_dependency(bash_job, parents=[parent_dependency.job])
+            for out_file in parent_dependency.output_files:
+                bash_job.add_inputs(out_file)
+
+        # Handle Output Files
```

Comment:
Lines 869-885 seem to be a duplication of code from the `run_container_func`. Can you possibly refactor into a method?

Reply:
Extracted the easily shared code which applies a variety of settings to jobs between the Bash, Container, and Python functions into a shared internal method.